### ISSUE TITLE
Issue in Generating release notes

### ISSUE THREAD

**Issue by nilukas** [rfennell/AzurePipelines#717]
I'm trying to generate the release notes using a release. I created a new release pipeline.
I have not created any tags in the work items. I committed the code to master branch adding a workitem. but not file is generated. and I tried to create a branch from master and created a pull request to master branch with adding workitem. but file was not generated.

my template is like this:

#Release notes for release $defname
**Release Number**  : $($testreleasenotes.name)    
**Release completed** $("{0:dd/MM/yy HH:mm:ss}" -f [datetime]$testreleasenotes.modifiedOn) 

### Associated work items
@@WILOOP@@
* **${widetail.fields['System.WorkItemType']} #${widetail.id}** Assigned by: ${widetail.fields['System.AssignedTo']} - ${widetail.fields['System.Title']}
@@WILOOP@@

![artifact](https://user-images.githubusercontent.com/10694532/82003262-8d10fa80-967d-11ea-9731-efe3e60a76a0.png)
![error](https://user-images.githubusercontent.com/10694532/82003264-8f735480-967d-11ea-918e-f6a8368a7b91.png)
![releasepipeline](https://user-images.githubusercontent.com/10694532/82003266-900beb00-967d-11ea-936a-0d436dd130cd.png)
![task1](https://user-images.githubusercontent.com/10694532/82003268-90a48180-967d-11ea-9bcd-7d33db7b40f7.png)
![task2](https://user-images.githubusercontent.com/10694532/82003269-913d1800-967d-11ea-9aae-2604930b355c.png)
![task3](https://user-images.githubusercontent.com/10694532/82003271-91d5ae80-967d-11ea-8df0-57fd7aa14b98.png)

please help me to resolve the issue

**Comment by rfennell** [comment]
The screenshot of the log says it has generated a file, with no associate WI/CS/Builds, but it should at least have a header with the release details.

A few things
- Set a variable `system.debug=true` this will provide more detailed logs for my task and any other
- Check the log for the copy task to make sure it is finding copying you file to thelocation you expect
- Given the working folder is `d:\a\_work\r1` am I right in assuming this is a Microsoft hosted agent? If so why are you copying the release notes file to the agent C: drive, when you have no access to this.?  You need to copy the file to a location off the agent e.g. send an email, upload to a WIKI etc. 
- An alternative is to try on a private agent, you will have access to the generate file without an upload

Let me know how it goes

**Comment by nilukas** [comment]
Thanks for the reply.
My requirement is
every developer is working in their own feature/dev branch and send the pull request to release branch for that month ex: rel2020.05. But they are not used to associate the work item to the pull request or their development branch, when they send the pull request. they miss it. So at the end of every month, I want to merge release branch to master and give the release to customer. When I give the release to customer I want to prepare the release note. That should include all completed work items  that developers work for that month. but please keep in mind, I have no commits or PRs asscociated with work items. I manually copying all wok items to atlassian wiki/confluence page. I want to automate this process.
You think header should print. but where should I add the reference for "testreleasenotes" in $testreleasenotes.name? Is that inbuilt name or a tag or anything else? I have not added that(testreleasenotes) to any where as a reference.
My expectation of copy task is to copy the file from $(System.DefaultWorkingDirectory)/ReleaseNotes.md to my local c drive, to make sure the file is generated with data.please suggest me once the file is generated with data, how can I see the data. 
copy file log is [2020-05-15T16:19:39.3713748Z found 0 files]. but no different
I added variable and trigger the release 
![variable](https://user-images.githubusercontent.com/10694532/82076324-5083e400-96fb-11ea-9d5b-c861d8bfb39f.png)

**Comment by rfennell** [comment]
I don't think these tasks work the way you think.

My release notes tasks uses the Azure DevOps api to find work items and commits associated with a build and hence a release. If your developers don't make work item associations with their commits you are not going to get any auto generated notes.

I don't think it cannot replace a manual process where you copy from jira. You need to get the developers to associate work items. So an Azure Devops  usage change for your team

As to copying files to your local disk, the copy task only copies between locations on the build host. It could only copy a location on your local disk if the agents is running on your local PC

**Comment by nilukas** [comment]
Thank you for your reply. I associated some wok items with commits and pull requests to make this work done. Even those work items, I could not generate the release notes, what's wrong with it? Please help me on creating release notes with your release notes behavior. Please verify my release artifacts and properties.and how I see the results or the file from my local machine without sending email?

**Comment by rfennell** [comment]
I have had another look at the screenshot of the non-debug log you sent  and notice that it says the `artifact type [Git]`. 

So am I right in assuming that the only artifact your release is associated with a a Git repo?

If so this task will not do what you require. It can only generate release notes based on an Azure DevOps Pipeline builds (either UI or YAML based) being run. This is because you have to run a build to get the basic data required. This task looks for changes in associated commits and work items BETWEEN BUILDS, in exactly the same way as the Azure DevOps UI. So if the Azure DevOps release UI shows no associations neither will my task.

Bottom line is it can get no commit or WI details if associated only with a Git or TFVC repo.

**Comment by wesleycamargo** [comment]
Hello, @rfennell I'm facing a similar problem.

I understand your point about we need related WI to generate a Release Notes.

But in my case, I have several artifacts, and some of them don't have this association. I'm trying to generate  HTML but it prints a lot of 'None' in the body:

![image](https://user-images.githubusercontent.com/897126/82370917-b7511800-99ef-11ea-9e27-3b6e6fe122ba.png)

I'm trying to find a way to skip the loop when it's empty, maybe it can be an option, but if you have some idea to solve this I'll appreciate it.

**Comment by rfennell** [comment]
@wesleycamargo with the legacy template you can't skip rows in the loop. However in the advanced option you can alter the text for empty rows from 'none' to something else including an empty string, do nothing is printed.

You have more options if you use the newer handlebars based templates as this allows if statement to be put inside the loops to give more control

**Comment by wesleycamargo** [comment]
@rfennell I followed your suggestion to change 'None' for empty text and it works fine.

Thank you and congratulations on this excellent task!

**Comment by rfennell** [comment]
As per my standard practice I am closing this issue as I have heard nothing for a while. If this issue remains a problem please do not hesitate to re-open it
